### ISSUE TITLE
Training Results Fixed When Using YOLOv8s and Larger Models with Custom IOU Loss Function

### ISSUE THREAD

**Issue by FoxyCarrot** [ultralytics/ultralytics#15867]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

#### Description:
I have observed an issue while training models using YOLOv8s and larger versions of YOLOv8. When I replace the default IOU loss function with a custom IOU loss function for YOLOv8s and models larger than YOLOv8s, the training results are fixed and do not exhibit any randomness across epochs. This fixed result is not observed when using YOLOv8n with the same custom IOU loss function and identical hyperparameters.

#### Dataset:
The dataset used is TWHD: Two Wheeler Helmet Dataset.

#### Question:
Could you please help me understand why the training results are fixed for YOLOv8s and larger models with the custom IOU loss function, while YOLOv8n does not exhibit this issue? Are there specific considerations or potential issues with the custom IOU loss function implementation that might affect YOLOv8s and larger models differently?

#### Additional Information:
- **Hyperparameters**: All other hyperparameters are kept the same.
- **Environment**: [Include details about your training environment if relevant, such as versions of software, hardware specifications, etc.]

Thank you for your assistance!

### Hyperparameters
All other hyperparameters are kept the same.

#### GPU Information

+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.171.04 Driver Version: 535.171.04 CUDA Version: 12.2 |
|-----------------------------------------+----------------------+----------------------+
| GPU Name Persistence-M | Bus-Id Disp.A | Volatile Uncorr. ECC |
| Fan Temp Perf Pwr:Usage/Cap | Memory-Usage | GPU-Util Compute M. |
| | | | | | MIG M. |
|=========================================+======================+======================|
| 0 NVIDIA GeForce RTX 4090 On | 00000000:13:00.0 Off | Off |
| 0% 34C P8 11W / 450W | 0MiB / 24564MiB | 0% Default |
| | | | | N/A |
+-----------------------------------------+----------------------+----------------------+

#### Installed Packages

absl-py==1.0.0
argon2-cffi==21.2.0
argon2-cffi-bindings==21.2.0
attrs==21.2.0
backcall==0.2.0
bleach==4.1.0
cachetools==5.0.0
catboost==1.0.3
certifi==2021.10.8
cffi==1.15.0
charset-normalizer==2.0.12
cloudpickle==2.0.0
cycler==0.11.0
debugpy==1.5.1
decorator==5.1.0
defusedxml==0.7.1
entrypoints==0.3
fonttools==4.28.3
google-auth==2.6.2
google-auth-oauthlib==0.4.6
graphviz==0.19
grpcio==1.42.0
h5py==3.6.0
idna==3.3
importlib-metadata==4.11.3
importlib-resources==5.4.0
ipykernel==6.6.0
ipympl==0.8.2
ipython==7.30.1
ipython-genutils==0.2.0
ipywidgets==7.6.5
jedi==0.18.1
Jinja2==3.0.3
joblib==1.1.0
jsonschema==4.2.1
jupyter-client==7.1.0
jupyter-core==4.9.1
jupyterlab-pygments==0.1.2
jupyterlab-widgets==1.0.2
kiwisolver==1.3.2
lightgbm==3.3.1
Markdown==3.3.6
MarkupSafe==2.0.1
matplotlib==3.5.0
matplotlib-inline==0.1.3
mistune==0.8.4
nbclient==0.5.9
nbconvert==6.3.0
nbformat==5.1.3
nest-asyncio==1.5.4
notebook==6.4.6
numpy==1.24.4
oauthlib==3.2.0
opencv-python==4.10.0.84
packaging==21.3
pandas==1.3.4
pandocfilters==1.5.0
parso==0.8.3
pexpect==4.8.0
pickleshare==0.7.5
Pillow==8.4.0
plotly==5.4.0
prometheus-client==0.12.0
prompt-toolkit==3.0.24
protobuf==3.20.0
psutil==5.9.0
ptyprocess==0.7.0
py-cpuinfo==9.0.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycocotools==2.0.4
pycparser==2.21
Pygments==2.10.0
pyparsing==3.0.6
pyrsistent==0.18.0
python-dateutil==2.8.2
pytz==2021.3
PyYAML==6.0
pyzmq==22.3.0
requests==2.27.1
requests-oauthlib==1.3.1
rsa==4.8
scikit-learn==1.0.1
scipy==1.10.1
seaborn==0.11.2
Send2Trash==1.8.0
setuptools-scm==6.3.2
six==1.16.0
sklearn==0.0
spyder-kernels==2.2.0
tenacity==8.0.1
tensorboard==2.8.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
terminado==0.12.1
testpath==0.5.0
threadpoolctl==3.0.0
tomli==1.2.2
torch==1.10.0+cu113
torchaudio==0.10.0+cu113
torchvision==0.11.1+cu113
tornado==6.1
tqdm==4.66.5
traitlets==5.1.1
typing_extensions==4.0.1
ultralytics==8.2.82
ultralytics-thop==2.0.5
urllib3==1.26.9
wcwidth==0.2.5
webencodings==0.5.1
Werkzeug==2.1.1
widgetsnbextension==3.5.2
wurlitzer==3.0.2
xgboost==1.5.1
zipp==3.6.0

**Comment by Y-T-G** [comment]
What do you mean fixed? They're the same across all epochs?

Or do you mean they are the same for all training runs?

**Comment by FoxyCarrot** [comment]
> What do you mean fixed? They're the same across all epochs?
> 
> Or do you mean they are the same for all training runs?

---

In the case of unchanged parameters, I train the same output for each epoch. Even if I change the parameters, it only results in another fixed outcome without randomness.But I clearly modified the IOU loss, and this phenomenon did not occur on YOLOV8n.

For example, when using YOLOV8S, running `yolo train model=yolov8s.pt data=helmet.yaml epochs=200 imgsz=640 batch=64`, with other parameters as default, the training result for each epoch must be:

| epoch | train/box_loss | train/cls_loss | train/dfl_loss | metrics/precision(B) | metrics/recall(B) | metrics/mAP50(B) | metrics/mAP50-95(B) | val/box_loss | val/cls_loss | val/dfl_loss | lr/pg0 | lr/pg1 | lr/pg2 |
|-------|----------------|----------------|----------------|-----------------------|-------------------|------------------|----------------------|--------------|--------------|--------------|---------|---------|---------|
| 1 | 1.246 | 1.4489 | 1.1249 | 0.5877 | 0.609 | 0.62354 | 0.37881 | 1.0057 | 0.85395 | 0.98894 | 0.003285| 0.003285| 0.003285|
| 2 | 1.0231 | 0.75649 | 0.97235 | 0.68517 | 0.64493 | 0.66643 | 0.39168 | 1.0752 | 0.81437 | 1.013 | 0.0065856| 0.0065856| 0.0065856|
| 3 | 1.0518 | 0.74484 | 0.97859 | 0.67509 | 0.65363 | 0.67449 | 0.39141 | 1.0927 | 0.97118 | 1.0297 | 0.0098532| 0.0098532| 0.0098532|
| 4 | 1.0643 | 0.73939 | 0.98588 | 0.71735 | 0.61858 | 0.65518 | 0.37394 | 1.1592 | 1.0307 | 1.0754 | 0.0098515| 0.0098515| 0.0098515|
| 5 | 1.0468 | 0.71417 | 0.98045 | 0.6614 | 0.61743 | 0.62586 | 0.36014 | 1.1975 | 1.0713 | 1.0755 | 0.009802 | 0.009802 | 0.009802 |
| 6 | 1.0268 | 0.69663 | 0.9751 | 0.7488 | 0.65256 | 0.70762 | 0.41412 | 1.1014 | 0.78255 | 1.0295 | 0.0097525| 0.0097525| 0.0097525|
| 7 | 1.0025 | 0.67637 | 0.96477 | 0.73482 | 0.68432 | 0.71457 | 0.41849 | 1.0698 | 0.89633 | 1.0132 | 0.009703 | 0.009703 | 0.009703 |
| 8 | 0.99758 | 0.65955 | 0.96413 | 0.73863 | 0.68954 | 0.73074 | 0.43232 | 1.0778 | 0.73473 | 1.0197 | 0.0096535| 0.0096535| 0.0096535|
| 9 | 0.98697 | 0.65004 | 0.95972 | 0.75734 | 0.70697 | 0.75217 | 0.44102 | 1.0623 | 0.699 | 1.0195 | 0.009604 | 0.009604 | 0.009604 |
| 10 | 0.97558 | 0.63963 | 0.95708 | 0.75278 | 0.69131 | 0.73471 | 0.43444 | 1.053 | 0.71354 | 1.0096 | 0.0095545| 0.0095545| 0.0095545|
| 11 | 0.96799 | 0.62401 | 0.95271 | 0.77073 | 0.72191 | 0.76611 | 0.45202 | 1.0102 | 0.67865 | 0.99071 | 0.009505 | 0.009505 | 0.009505 |
| 12 | 0.95354 | 0.61817 | 0.94843 | 0.76284 | 0.71817 | 0.75513 | 0.4582 | 0.99974 | 0.6543 | 0.98595 | 0.0094555| 0.0094555| 0.0094555|
| 13 | 0.94317 | 0.60489 | 0.94432 | 0.77045 | 0.71015 | 0.75971 | 0.45294 | 1.0205 | 0.6761 | 0.98942 | 0.009406 | 0.009406 | 0.009406 |
| 14 | 0.93811 | 0.60018 | 0.94161 | 0.7562 | 0.74323 | 0.77041 | 0.46317 | 0.99906 | 0.65884 | 0.97484 | 0.0093565| 0.0093565| 0.0093565|
| 15 | 0.92338 | 0.58386 | 0.93782 | 0.79411 | 0.71603 | 0.77256 | 0.46483 | 0.99796 | 0.64443 | 0.98027 | 0.009307 | 0.009307 | 0.009307 |
| 16 | 0.92667 | 0.582 | 0.93518 | 0.77042 | 0.72457 | 0.77551 | 0.47055 | 0.98419 | 0.63008 | 0.98148 | 0.0092575| 0.0092575| 0.0092575|
| 17 | 0.91665 | 0.57485 | 0.93559 | 0.78319 | 0.72801 | 0.7815 | 0.4712 | 0.99489 | 0.63652 | 0.98232 | 0.009208 | 0.009208 | 0.009208 |
| 18 | 0.90993 | 0.57254 | 0.93184 | 0.7808 | 0.73863 | 0.78594 | 0.48222 | 0.97942 | 0.61296 | 0.9799 | 0.0091585| 0.0091585| 0.0091585|
| 19 | 0.90493 | 0.56252 | 0.92951 | 0.80753 | 0.73232 | 0.78947 | 0.48122 | 0.97653 | 0.60596 | 0.97289 | 0.009109 | 0.009109 | 0.009109 |
| 20 | 0.90111 | 0.56143 | 0.92865

**Comment by glenn-jocher** [comment]
@FoxyCarrot thank you for the detailed explanation. It seems like the issue might be related to the implementation of your custom IOU loss function. Larger models like YOLOv8s may be more sensitive to certain aspects of the loss function, leading to the fixed results you're observing. I recommend reviewing the custom IOU loss function for any potential issues or differences in behavior when applied to larger models. Additionally, ensure that the loss function is properly integrated and that there are no unintended side effects. If the issue persists, please verify with the latest version of the Ultralytics package.
